src/algorithms/semihdp_sampler.cpp

```diff
@@ -0,0 +1,713 @@
+      s[i][j] = bayesmix::categorical_rng(probas, rng);
+    }
+
+    v[i].resize(2 * INIT_N_CLUS);
+    t[i].resize(2 * INIT_N_CLUS);
+
+    for (int l = 0; l < INIT_N_CLUS; l++) {
+      std::shared_ptr<BaseHierarchy> hierarchy = G0_master_hierarchy->clone();
+      hierarchy->draw();
+      theta_tilde[i].push_back(hierarchy);
+      theta_star[i].push_back(hierarchy);
+      v[i][l] = l;
+      t[i][l] = -1;
+    }
+
+    for (int l = 0; l < INIT_N_CLUS; l++) {
+      theta_star[i].push_back(taus[l]);
+      v[i][INIT_N_CLUS + l] = -1;
+      t[i][INIT_N_CLUS + l] = l;
+    }
+  }
+  m = std::vector<int>(taus.size(), 0);
+
+  for (int i = 0; i < ngroups; i++) {
+    n_by_theta_star[i] = std::vector<int>(theta_star[i].size(), 0);
+    for (int j = 0; j < n_by_group[i]; j++) {
+      n_by_theta_star[i][s[i][j]] += 1;
+    }
+  }
+}
+
+void SemiHdpSampler::update_unique_vals() {
+  // Loop through the theta stars, if its equal to some theta_tilde, we
+  // update it on the fly. Otherwise we store the the data in a vector
+  // and then update all the taus
+
+  std::vector<MatrixXd> data_by_tau(taus.size());
+  for (int h = 0; h < taus.size(); h++) {
+    data_by_tau[h] = Eigen::MatrixXd(0, 1);
+  }
+
+  for (int r = 0; r < ngroups; r++) {
+    std::vector<MatrixXd> data_by_theta_star(theta_star[r].size());
+    for (int i = 0; i < ngroups; i++) {
+      if (c[i] == r) {
+        for (int j = 0; j < n_by_group[i]; j++) {
+          bayesmix::append_by_row(&data_by_theta_star[s[i][j]],
+                                  data[i].row(j));
+        }
+      }
+    }
+
+    for (int l = 0; l < theta_star[r].size(); l++) {
+      if (v[r][l] >= 0) {
+        if (data_by_theta_star[l].rows() > 0)
+          theta_tilde[r][v[r][l]]->sample_given_data(data_by_theta_star[l]);
+        else
+          theta_tilde[r][v[r][l]]->draw();
+        theta_star[r][l] = theta_tilde[r][v[r][l]];
+      } else {
+        bayesmix::append_by_row(&data_by_tau[t[r][l]], data_by_theta_star[l]);
+      }
+    }
+  }
+
+  for (int h = 0; h < taus.size(); h++) {
+    if (data_by_tau[h].rows() > 0)
+      taus[h]->sample_given_data(data_by_tau[h]);
+    else
+      taus[h]->draw();
+  }
+
+  // reassign stuff to teta_stars
+  for (int i = 0; i < ngroups; i++) {
+    for (int l = 0; l < theta_star[i].size(); l++) {
+      if (t[i][l] >= 0) {
+        theta_star[i][l] = taus[t[i][l]];
+      }
+    }
+  }
+}
+
+void SemiHdpSampler::update_s() {
+
+  auto& rng = bayesmix::Rng::Instance().get();
+  const double logw = std::log(w);
+  const double log1mw = std::log(1 - w);
+  const double logalpha = std::log(alpha);
+  const double loggamma = std::log(gamma);
+
+  // compute counts
+  _count_m();
+  _count_n_by_theta_star();
+
+  std::vector<std::vector<int>> log_n_by_theta_star(ngroups);
+  for (int r = 0; r < ngroups; r++) {
+    for (int n : n_by_theta_star[r]) {
+      if (n == 0)
+        log_n_by_theta_star[r].push_back(1e-20);
```

Comment:
I've seen you doing this elsewhere: why do you push 1e-20 instead of 0 or 0.0?

Reply:
Thanks, I worried a bit of logarithm instability when using some optimization flags such as fast-math, but seems to be fine!